File: leetcode-scala/src/dfs/medium/ConvertSortedListToBinarySearchTree.scala
package dfs.medium

import tools.{ListNode, TreeNode}

/**
 * @Project leetcode
 * @Author Ravooo
 * @Since 2020/7/16 10:24
 * @Description ConvertSortedListToBinarySearchTree
 */
object ConvertSortedListToBinarySearchTree {

    /**
     * 2744 ms  33.33%
     * 58 MB    100.00%
     * 将ListNode转为升序数组 再根据数组构建二叉搜索树
     * @param head
     * @return
     */
    def sortedListToBST(head: ListNode): TreeNode = {
        var elements = Array[Int]()
        var cur = head
        Iterator.continually(cur).takeWhile(_ != null).foreach(t => {
            elements :+= t.x
            cur = cur.next
        })

        def constructBSTFromArray(elements: Array[Int]): TreeNode = {
            if (elements == null || elements.length == 0) return null
            if (elements.length == 1) return new TreeNode(elements(0))

            val mid = elements.length / 2
            val root = new TreeNode(elements(mid))
            if (mid > 0) {
                root.left = constructBSTFromArray(elements.slice(0, mid))
            }
            if (mid < elements.length) {
                root.right = constructBSTFromArray(elements.slice(mid + 1, elements.length))
            }

            root
        }

        constructBSTFromArray(elements)
    }
}

File: leetcode-java/src/classify/binary_search_tree/medium/ConvertSortedListToBinarySearchTree.java
package classify.binary_search_tree.medium;

import tools.ListNode;
import tools.TreeNode;

public class ConvertSortedListToBinarySearchTree {
    
    /**
     * 0ms      100.00%
     * 41.9MB   99.37%
     */
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return new TreeNode(head.val);

        ListNode slowPre = null, slow = head, fast = head.next;
        while (fast != null && fast.next != null) {
            slowPre = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        TreeNode node = new TreeNode(slow.val);
        if (slowPre != null) slowPre.next = null;
        node.right = sortedListToBST(slow.next);
        if (slow != head) {
            node.left = sortedListToBST(head);
        }

        return node;
    }
}
